### ISSUE TITLE
How to handle AdditionalProperties

### ISSUE THREAD

**Issue by TristanSpeakEasy** [pb33f/libopenapi#29]
Looking into the code AdditionalProperties is an any type which seems to be populated directly from the yaml node.

But it can be either a bool or a schema according to spec.

If its a schema the code currently returns this as a map[string]any I believe, how do I resolve this to the schema type?

**Comment by TristanSpeakEasy** [comment]
So for an object with an AdditionalProperties attribute it can either be defined as:

```yaml
schema:
  type: object
  additionalProperties: true
```

or

```yaml
schema:
  type: object
  additionalProperties:
    type: string
```

here is some code I have trying to handle this, it gets an additionalProperties map but that actually represents a schema

```go
        var additionalProperties map[string]any
	additionalPropertiesAllowed := false
	if g.getOpenAPIVersion() == openAPIVersion31 {
		additionalPropertiesAllowed = true
	}

	apAllowed, ok := schema.AdditionalProperties.(bool)
	if ok {
		additionalPropertiesAllowed = apAllowed
	} else {
		ap, ok := schema.AdditionalProperties.(map[string]any)
		if ok {
			additionalProperties = ap
		}
	}
```

(btw would be great to have a helper function to get the exact type of the document as an enum or similar for handling 3 vs 3.1 differences efficiently)

**Comment by TristanSpeakEasy** [comment]
sorry it can also be an empty map as well:

```
schema:
  type: object
  additionalProperties: {}
```

which is equivalent to `additionalProperties: true`

in 3.1 btw additionalProperties defaults to true if not set in the document

**Comment by daveshanley** [comment]
Hmm, OK, I see. I'll noodle on this and think about a clean way to handle it.

**Comment by danielgtaylor** [comment]
This is how I'm handling `AdditionalProperties` and it seems to work. How are you getting a `map[string]any`? My test input is:

```yaml
{type: object, additionalProperties: {type: string}}
```

Code:

```go
if s.AdditionalProperties != nil {
	ap := s.AdditionalProperties
	if sp, ok := ap.(*lowbase.SchemaProxy); ok {
		ap = sp.Schema()
	}
	if low, ok := ap.(*lowbase.Schema); ok {
		addl := base.NewSchema(low)
		simple := isSimpleSchema(addl)
		hash := low.Hash()
		if simple || !known[hash] {
			known[hash] = true
			obj += indent + "  " + "<any>: " + renderSchemaInternal(addl, indent+"  ", mode, known) + "\n"
		} else {
			obj += indent + "  <any>: <rescurive ref>\n"
		}
	}
	if b, ok := ap.(bool); ok && b {
		obj += indent + "  <any>: <any>\n"
	}
}
```

https://github.com/danielgtaylor/restish/pull/145/files#diff-31fa877c3811387c2e44efc5df64497854208a9ba40c8e33e7c52b578a850f2fR193-R210

**Comment by TristanSpeakEasy** [comment]
Thanks Daniel for the example code, this was largely based on an assumption by me as I am retrofitting the library into existing code but haven't got it to a compilable stage yet so largely trying to deduce handling of various things through looking at the code.

If it does work as in your example then that is great and I will give that a go.

But interested in how it handles 

```
{type: object, additionalProperties: {}}
```

**Comment by daveshanley** [comment]
There is currently no relationship between a `SchemaProxy` and a `Schema,` only the other way around, it's a one-way direction, however, if it makes sense, it won't be hard to add a pointer back.

**Comment by TristanSpeakEasy** [comment]
Is it building the schema previously that can cause this to be a Schema at some point? Could it always just be a schema proxy that the schema needs to be retrieved from? Otherwise some link back to the proxy would be great as then we can stick to using a schemaproxy as the type we pass in args.

Basically our use case is we are recursively drilling down the chain of schemas and building an AST for code generation with the object currently being passed from recursive call to recursive call being the schemaproxy.

**Comment by daveshanley** [comment]
The `SchemaProxy` acts as a break-glass to stop circular errors. It hibernates until a `Schema()` call is made, the underlying `Schema` is then built and then attached to the `SchemaProxy` (so it won't be re-built with any subsequent calls to `Schema()`) and then is returned. 

Any further calls to `Schema()` just return a pointer to the already-built schema. 

The `Schema` has no idea which `SchemaProxy` holds its reference. It's trivial to add a property called `ParentProxy` (or something) of type `*SchemaProxy`

This is something that will be set when the schema is built and will provide a link back to the `*SchemaProxy`. Would this design work for you?

**Comment by TristanSpeakEasy** [comment]
Yep that would work 👍

**Comment by daveshanley** [comment]
Done! available in `v0.3.6`

**Comment by TristanSpeakEasy** [comment]
@daveshanley having been debugging an issue where I haven't been able to cast AdditionalProperties to `*base.SchemaProxy` and I discovered it is because AdditionalProperties  is a "low" `*base.SchemaProxy` as opposed to a "high" `*base.SchemaProxy` while every other property in the schema uses a high schema proxy I would expect this one to as well.

Is this a bug or expected behaviour?

**Comment by daveshanley** [comment]
This is a bug; there should be no mixing of high and low props, most likely an import issue. I will address ASAP.

**Comment by daveshanley** [comment]
I'm looking at the code at the moment, the `AdditionalProperties` prop in the high-level is just passing through the `interface{}` value and not the low `SchemaProxy`. 

Do you have an example of the code/spec? I won't be able to reproduce the issue correctly without it

**Comment by TristanSpeakEasy** [comment]
Here is a code snippet that reproduces the issue

```go
package main

import (
	"fmt"
	"os"
	"reflect"

	"github.com/pb33f/libopenapi"
	"github.com/pb33f/libopenapi/datamodel/high/base"
)

func main() {
	data, err := os.ReadFile("./openapi.yaml")
	if err != nil {
		panic(err)
	}

	doc, err := libopenapi.NewDocument(data)
	if err != nil {
		panic(err)
	}

	model, errs := doc.BuildV3Model()
	if len(errs) > 0 {
		panic(errs)
	}

	ops := model.Model.Paths.PathItems["/v1/apis"].GetOperations()

	schema, err := ops["get"].Responses.Codes["200"].Content["application/json"].Schema.BuildSchema()
	if err != nil {
		panic(err)
	}

	itemsSchema, err := schema.Items.A.BuildSchema()
	if err != nil {
		panic(err)
	}

	metadataSchema, err := itemsSchema.Properties["meta_data"].BuildSchema()
	if err != nil {
		panic(err)
	}

	fmt.Println(reflect.TypeOf(metadataSchema.AdditionalProperties))

	ap, ok := metadataSchema.AdditionalProperties.(*base.SchemaProxy)
	if !ok {
		panic("not a high schema proxy")
	}

	fmt.Println(reflect.TypeOf(ap))
}
```

using our OpenAPI doc you can grab here: https://docs.speakeasyapi.dev/openapi.yaml

**Comment by daveshanley** [comment]
Thank you!

This was 100% a bug, and is now fixed in `v0.4.1`

**Comment by TristanSpeakEasy** [comment]
Can confirm it is fixed
